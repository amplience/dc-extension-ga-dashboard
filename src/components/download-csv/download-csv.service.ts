import { json2csvAsync } from 'json-2-csv';

export async function generateCsvBlob(data: unknown): Promise<Blob> {
  const csv = await json2csvAsync(JSON.parse(JSON.stringify(data)));
  return new Blob([csv]);
}

export function createDownloadElement(
  blob: Blob,
  fileName: string
): HTMLAnchorElement {
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.setAttribute('href', url);
  link.setAttribute('download', fileName);
  link.style.visibility = 'hidden';
  return link;
}
