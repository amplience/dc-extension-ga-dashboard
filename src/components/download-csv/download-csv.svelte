<script lang="ts">
  import {
    generateCsvBlob,
    createDownloadElement,
  } from './download-csv.service';
  import Icon from '../icon/icon.svelte';
  import DownloadIcon from '../../assets/icons/ic-download.svg';
  import { createEventDispatcher } from 'svelte';

  export let label: string = '';
  export let fileName: string;
  export let fetchCsvData: () => unknown;

  const dispatch = createEventDispatcher();

  async function onClick() {
    const data = await fetchCsvData();
    const csvBlob = await generateCsvBlob(data);
    let csvFilename = `${fileName}.csv`;
    const downloadElement = createDownloadElement(csvBlob, csvFilename);
    document.body.appendChild(downloadElement);
    downloadElement.click();
    document.body.removeChild(downloadElement);

    dispatch('downloadCsv', {
      filename: csvFilename,
    });
  }
</script>

<style>
  button {
    cursor: pointer;
    border: none;
    background: none;
    padding: 0;
    margin: 0;
    display: flex;
    align-items: center;
  }

  button :global(svg) {
    width: 16px;
    height: 20px;
    color: #666;
  }

  button span {
    color: #039be5;
    margin-right: 4px;
  }
</style>

<button
  on:click={onClick}
  class="download-csv-button"
  data-testid="download-csv-button"
  title="Download CSV (max 1000)">
  {#if label}
    <span>{label}</span>
  {/if}
  <Icon icon={DownloadIcon} />
</button>
