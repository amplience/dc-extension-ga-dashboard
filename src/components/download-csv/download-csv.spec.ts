import { render, fireEvent, screen } from '@testing-library/svelte';
import DownloadCsv from './download-csv.svelte';
import { tick } from 'svelte';
import { generateCsvBlob, createDownloadElement } from './download-csv.service';

jest.mock('./download-csv.service');

describe('DownloadCsv', () => {
  it('should render the DownloadCsv component', () => {
    const { container } = render(DownloadCsv, {
      fileName: 'test-csv-filename',
      fetchCsvData: () => {
        /* noop */
      },
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the DownloadCsv component with a label', () => {
    const { container } = render(DownloadCsv, {
      fileName: 'test-csv-filename',
      label: 'test-label',
      fetchCsvData: () => {
        /* noop */
      },
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should trigger a csv download on click', async () => {
    (generateCsvBlob as jest.Mock).mockImplementation(
      () => new Blob(['test,blob'])
    );
    const clickMock = jest.fn();
    (createDownloadElement as jest.Mock).mockImplementation(() => {
      const anchor = document.createElement('a');
      anchor.click = clickMock;
      return anchor;
    });
    const fetchCsvDataMock = jest
      .fn()
      .mockImplementation(() => ({ testProp: 'test prop value' }));

    jest.spyOn(document.body, 'appendChild');
    jest.spyOn(document.body, 'removeChild');

    render(DownloadCsv, {
      fileName: 'test-csv-filename',
      fetchCsvData: fetchCsvDataMock,
    });

    const element = screen.getByTestId('download-csv-button');

    await fireEvent.click(element);
    await tick();

    expect(fetchCsvDataMock).toHaveBeenCalled();
    expect(generateCsvBlob).toHaveBeenCalled();
    expect(createDownloadElement).toHaveBeenCalledWith(
      expect.any(Blob),
      'test-csv-filename.csv'
    );
    expect(document.body.appendChild).toHaveBeenCalled();
    expect(clickMock).toHaveBeenCalled();
    expect(document.body.removeChild).toHaveBeenCalled();
  });
});
