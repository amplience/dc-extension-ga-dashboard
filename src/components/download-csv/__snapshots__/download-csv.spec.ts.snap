// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCsv should render the DownloadCsv component 1`] = `
<div>
  <button
    class="download-csv-button svelte-wktmji"
    data-testid="download-csv-button"
    title="Download CSV (max 1000)"
  >
     
    <div
      class=" svelte-zpbvvv"
      style="--icon-width: 24px; --icon-height: 24px;"
    >
      <object
        src="ic-download.svg/"
      />
    </div>
  </button>
</div>
`;

exports[`DownloadCsv should render the DownloadCsv component with a label 1`] = `
<div>
  <button
    class="download-csv-button svelte-wktmji"
    data-testid="download-csv-button"
    title="Download CSV (max 1000)"
  >
    <span
      class="svelte-wktmji"
    >
      test-label
    </span>
     
    <div
      class=" svelte-zpbvvv"
      style="--icon-width: 24px; --icon-height: 24px;"
    >
      <object
        src="ic-download.svg/"
      />
    </div>
  </button>
</div>
`;
